exporter/elasticsearchexporter/internal/objmodel/objmodel.go

```diff
@@ -283,11 +283,15 @@ func (doc *Document) iterJSONFlat(w *json.Visitor, otel bool) error {
 	return nil
 }
 
-// Set of prefixes for the OTel attributes that needs to stay flattened
+// Under OTel mode, set of key prefixes where keys should be flattened from that level.
+// It is not aware of whether it is a root document or sub-document.
+// NOTE: This works very delicately with the implementation of OTel mode that
+// e.g. resource.attributes is a "resource" objmodel.Document under the root document that contains attributes
+// added using AddAttributes func as flattened keys.
+// Therefore, there will be correctness issues when attributes are added / used in other ways, but it is working
+// for current use cases and the proper fix will be slightly too complex. YAGNI.
 var otelPrefixSet = map[string]struct{}{
-	"attributes.":          {},
-	"resource.attributes.": {},
-	"scope.attributes.":    {},
+	"attributes.": {},
```

Comment:
I have not followed why this was introduced so I might be missing something but what is the need to have this `otelPrefixSet`? IIUC, `dedotting` is only for ECS mode so this method should not be called at all for otel mode, is this supposed to be temporary until dedot config is available in the exporter?

Reply:
This is to ensure the following happens to a document:
```
attributes.a.b=1
```
instead of becoming `{attributes: {a: {b: 1}}}` under dedot=true, it becomes `{attributes: {a.b:1}}` under dedot=true and otel=true.

> dedotting is only for ECS mode

While the README says so, dedotting is enabled for otel mode and it does need dedot except for a few fields, e.g. resource.attributes, scope.attributes and attributes.

Reply:
> dedotting is enabled for otel mode and it does need dedot except for a few fields

e.g. `resource.attributes.a.b=1` should be emitted as `{resource: {attributes: {a.b:1}}}` in otel mode.

Reply:
>  it does need dedot except for a few fields, e.g. resource.attributes, scope.attributes and attributes.

It this because of `passthrough`?

Reply:
> It this because of passthrough?

Correct

Reply:
> 	"resource.attributes.": {},
	"scope.attributes.":    {},

For my understanding, the reason why we are removing these 2 is because resource and scope attributes are added using [ValueFromAttributes](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/8c8ffb4bc17315b15e0993cd13b72c68e4e66321/exporter/opensearchexporter/internal/objmodel/objmodel.go#L384) and that already adds them as dotted fields which works for us?

Reply:
They are added using `AddAttribute`, not `ValueFromAttributes`, and that adds them as dotted fields with flattened keys. Then in iterJSONDedot, we dedot until the level of `attributes.`.

Reply:
Hmm, I am confused now, resource attributes are added as `"resource": {"attributes": ...}` using `ValueFromAttributes`: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3d94d51311a98d68ae034bb6b2b2d6175acec033/exporter/elasticsearchexporter/model.go#L187

They end up as dotted field from what I can see. The `attributes.` refer to the span attributes or dp attributes.

Reply:
`ValueFromAttributes` makes `resource` appear as a sub-document under `document`. Any maps under resource as added as flattened as they are added in turn using `AddAttribute` / `AddAttributes`.

If `document.Add("attributes", objmodel.ValueFromAttribute(attrMap))`, the `iterJSONDedot` func will fail to handle this, as `attributes` is a subdocument under root, then under attributes is the map, instead of a flattened map (with keys prefix `attributes.`) under root.

Reply:
Taked IRL. For posterity, we don't need `resource.attributes.` and `scope.attributes.` as they are added in expected format.